### PR TITLE
Feat/dflash training improvements

### PR DESCRIPTION
**Pull request by xiaomin-D**
## Motivation

The current DFlash training implementation in SpecForge diverges from the DFlash paper ([arXiv:2602.06036](https://arxiv.org/abs/2602.06036)) in several key aspects: it uses uniform block division instead of Random Anchor Sampling, lacks Loss Decay Weighting, and has misaligned hyperparameters. Additionally, the saved model config format does not match the upstream DFlash repository, causing loading incompatibilities.

This PR closes the gap between SpecForge's DFlash training and the paper/upstream, and improves training data efficiency and loss stability.

## Modifications

**Training strategies (paper Sec 4.2):**
- **Random Anchor Sampling**: Randomly sample anchor positions from the response region to construct training blocks, instead of uniform block division. Anchors are sampled independently per batch sample (max strategy) to maximize data utilization. Padding blocks are isolated via `block_id=-1` in attention mask to prevent loss instability.
- **Loss Decay Weighting** (paper Eq.4, Appendix A.3.1): Apply exponential decay `w_k = exp(-(k-1)/γ)` to per-token loss based on position within each block, reflecting the diminishing importance of later tokens in speculative decoding.

**Upstream sync:**
- Read `mask_token_id` and `target_layer_ids` from `config.dflash_config` (with fallback), matching upstream DFlash's model loading convention.
- Write `dflash_config` into saved checkpoints for HuggingFace compatibility.
- Fix `auto_map` entry and saved model file name (`dflash.py` instead of `modeling_dflash.py`).

**Hyperparameter alignment (paper Appendix A.1):**
- `lr=6e-4`, `warmup_ratio=0.04`, `num_epochs=6`, `max_length=3072`, `max_grad_norm=1.0`

**Files changed:**
- `specforge/core/dflash.py` — Core training logic: random anchor sampling, loss decay, per-sample block isolation
- `specforge/modeling/draft/dflash.py` — Read config from `dflash_config`
- `scripts/train_dflash.py` — New CLI args, hyperparameter defaults, checkpoint saving fix
- `configs/qwen3-8b-dflash.json`, `configs/longcat-flash-dflash.json` — Add `dflash_config`, fix `auto_map`
- `examples/run_qwen3_8b_dflash_online.sh`, `examples/run_longcat_flash_dflash_online.sh` — Updated example scripts

## Related Issues

N/A

## Accuracy Test

Pending — to be validated with full training runs on Qwen3-8B targets.

## Benchmark & Profiling

N/A (training-side changes only; inference path unchanged)

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://sgl-fru7574.slack.com/archives/C09784E3EN6 to discuss your PR.

### CODE REVIEW

**Comment by xiaomin-D** [comment]
Hi team, I've updated the training process based on the Dflash paper and am currently experimenting with it. Welcome to discuss it together～

**Comment by ggg-s** [comment]
hi @xiaomin-D  It seems that **`OnlineDFlashModel` does not include any handling related to `random_anchor`**.

**Comment by xiaomin-D** [comment]
> hi @xiaomin-D It seems that **`OnlineDFlashModel` does not include any handling related to `random_anchor`**.


Hi, thanks for the review! `OnlineDFlashModel` does handle `random_anchor` — it's controlled by the `random_anchor` flag passed to `__init__` (L33, L44). The core logic is in `forward()` at L259:

```python
if self.random_anchor and self.training:
    anchor_positions, block_keep_mask = self._sample_anchor_positions(...)
    (input_ids, hidden_states, loss_mask, block_ids, block_positions) = (
        self._build_blocks_from_anchors(...)
    )
```

When `random_anchor=True` and the model is in training mode, it calls `_sample_anchor_positions()` (L52) to randomly sample anchors per sample, then `_build_blocks_from_anchors()` (L90) to gather blocks. Padding blocks are isolated via `block_id=-1` in both attention mask and loss mask. When `random_anchor=False`, it falls through to the standard uniform block path (L270).

The related CLI arg is `--random-anchor` in `train_dflash.py`, and the example scripts (`run_qwen3_8b_dflash_online.sh`) enable it by default.

File: specforge/modeling/draft/dflash.py [hunk 1]
```diff
@@ -267,20 +270,18 @@ def spec_generate(
         self,
         target: nn.Module,
         input_ids: torch.LongTensor,
-        mask_token_id: int,
         max_new_tokens: int,
         stop_token_ids: list[int],
         temperature: float,
     ):
         self.eval()
```
[the remark this answers is not in the document]
Comment (xiaomin-D) [reply specforge/modeling/draft/dflash.py:277]: This is for alignment with the zlab/dflash code; the API was not called, so there is no risk.
